code/game/atoms.dm

```diff
@@ -362,40 +362,125 @@
 
-	blood_color = "#A10808"
-	if(istype(M))
-		if(M.species.flags & NO_BLOOD)
-			return 0
-		M.check_dna()
-		blood_color = M.species.blood_color
+/mob/living/carbon/get_blood_dna_list()
+	if(get_blood_id() != "blood")
+		return
+	var/list/blood_dna = list()
+	if(dna)
+		var/mob/living/carbon/human/H = src
+		blood_dna[dna.unique_enzymes] = H.b_type
+	else
+		blood_dna["UNKNOWN DNA"] = "X*"
+	return blood_dna
 
-	. = 1
-	return 1
+/mob/living/carbon/alien/get_blood_dna_list()
+	return list("UNKNOWN DNA" = "X*")
 
-/atom/proc/add_blood_list(mob/living/carbon/M)
-	// Returns 0 if we have that blood already
-	if(!istype(blood_DNA, /list))	//if our list of DNA doesn't exist yet (or isn't a list) initialise it.
+//to add a mob's dna info into an object's blood_DNA list.
+/atom/proc/transfer_mob_blood_dna(mob/living/L)
+	var/new_blood_dna = L.get_blood_dna_list()
+	if(!new_blood_dna)
+		return 0
+	return transfer_blood_dna(new_blood_dna)
+
+/obj/effect/decal/cleanable/blood/splatter/transfer_mob_blood_dna(mob/living/L)
+	..(L)
+	var/list/b_data = L.get_blood_data(L.get_blood_id())
+	basecolor = b_data["blood_color"]
+	update_icon()
+
+/obj/effect/decal/cleanable/blood/footprints/transfer_mob_blood_dna(mob/living/L)
+	..(L)
+	var/list/b_data = L.get_blood_data(L.get_blood_id())
+	basecolor = b_data["blood_color"]
+	update_icon()
+
+//to add blood dna info to the object's blood_DNA list
+/atom/proc/transfer_blood_dna(list/blood_dna)
+	if(!blood_DNA)
 		blood_DNA = list()
-	//if this blood isn't already in the list, add it
-	if(blood_DNA[M.dna.unique_enzymes])
-		return 0 //already bloodied with this blood. Cannot add more.
-	var/blood_type = "X*"
-	if(ishuman(M))
-		var/mob/living/carbon/human/H = M
-		blood_type = H.b_type
-	blood_DNA[M.dna.unique_enzymes] = blood_type
-	return 1
+	var/old_length = blood_DNA.len
+	blood_DNA |= blood_dna
+	if(blood_DNA.len > old_length)
+		return 1//some new blood DNA was added
 
-// Only adds blood on the floor -- Skie
-/atom/proc/add_blood_floor(mob/living/carbon/M)
-	return //why the fuck this is at an atom level but only works on simulated turfs I don't know
+
+//to add blood from a mob onto something, and transfer their dna info
+/atom/proc/add_mob_blood(mob/living/M)
+	var/list/blood_dna = M.get_blood_dna_list()
+	if(!blood_dna)
+		return 0
+	var/list/b_data = M.get_blood_data(M.get_blood_id())
+
+	return add_blood(blood_dna, b_data["blood_color"])
+
+//to add blood onto something, with blood dna info to include.
+/atom/proc/add_blood(list/blood_dna, color)
+	return 0
+
+/obj/add_blood(list/blood_dna, color)
+	return transfer_blood_dna(blood_dna)
+
+/obj/item/add_blood(list/blood_dna, color)
+	var/blood_count = !blood_DNA ? 0 : blood_DNA.len
+	if(!..())
+		return 0
+	if(!blood_count)//apply the blood-splatter overlay if it isn't already in there
+		add_blood_overlay(blood_DNA, color)
+	return 1 //we applied blood to the item
+
+/obj/item/proc/add_blood_overlay(list/blood_dna, color)

```

Comment:
Maybe bullets embedding won't lag the server to hell anymore (remember the fully automatic shotgun with buckshot? You could lag the server really badly just by blood overlay icon generation on the embedded bullets).